core/src/memory/default_resource_polyfill.cpp

```diff
@@ -0,0 +1,57 @@
+#include <atomic>
+
+#include "vecmem/memory/memory_resource.hpp"
+
+#if defined(VECMEM_HAVE_PMR_MEMORY_RESOURCE)
+namespace std::pmr {
+#else
+namespace std::experimental {
+inline namespace fundamentals_v1 {
+namespace pmr {
+#endif
+
+namespace {
+class new_delete_resource_impl : public memory_resource {
+    virtual void* do_allocate(std::size_t bytes, std::size_t) override {
+        return malloc(bytes);
+    }
+
+    virtual void do_deallocate(void* p, std::size_t, std::size_t) override {
+        free(p);
+    }
+
+    virtual bool do_is_equal(
+        const memory_resource& other) const noexcept override {
+        return &other == this;
+    }
+};
```

Comment:
Now that it's in a `.cpp` file, there's nothing preventing you from just using `vecmem::host_memory_resource`, right?

Reply:
Only `vecmem::host_memory_resource::do_equal`, which the spec says should return `this == &other`. `vecmem::host_memory_resource::do_equal` does something different, which I believe ultimately results in the same result boolean. If I supply this as `new_delete_resource`, I'd probably want it to match `this == &other`, up to you if you want me to modify `vecmem::host_memory_resource::do_equal` to do that, in which case I can remove `new_delete_resource_impl`.

Reply:
:thinking: I don't know... Yes, the documentation does say that the standard implementation of the object returned by [new_delete_resource()](https://en.cppreference.com/w/cpp/memory/new_delete_resource) should implement its `is_equal(...)` check through pointer comparison. But I fail to see that as a good design. :confused:

This comparison function to me meant that the runtime could decide with it whether some memory resources could cross-manage each others' resources. And two different instances of `vecmem::host_memory_resource` can very much do that. While two different instances of (for instance) `vecmem::cuda::device_memory_resource` not necessarily. So it makes (in my mind) perfect sense to allow non-trivial implementations for this comparison function.

I see the implementation logic in what the standard says. Since the standard does not define a public type that would manage memory through `new` / `delete`, it only provides the `new_delete_resource()` function, the pointer comparison is "fine". Since you are not supposed to be able to instantiate the underlying resource multiple times.

But I just hate that part of the design... :frowning: These helper functions make sense to have, to make sure that default constructed containers could "do something". But I don't agree with the part of the design that disallows creating "host memory resources" on the stack...

So long story short, I would really like to keep using `vecmem::host_memory_resource` as is, and provide a static instance of it through `new_delete_resource()`. (Is this latter also not provided by `libc++`? Did you check? Should we maybe provide these functions one-by-one, after having checked for their existence one-by-one?)

Reply:
I checked on macOS at least, and `new_delete_resource()` seems not to be supplied, so needs to be implemented.

Reply:
Done.